SourceX/controls/plrctrls.cpp

```diff
@@ -0,0 +1,689 @@
+
+namespace dvl {
+
+#define INV_TOP 240;
+#define INV_LEFT RIGHT_PANEL_X;
+#define INV_HEIGHT 320;
+
+bool newCurHidden = false;
+coords speedspellscoords[50];
+int speedspellcount = 0;
+
+// Native game menu, controlled by simulating a keyboard.
+bool InGameMenu()
+{
+	return stextflag > 0 || questlog || helpflag || talkflag || qtextflag || sgpCurrentMenu;
+}
+
+namespace {
+
+DWORD invmove = 0;
+int hsr[3] = { 0, 0, 0 }; // hot spell row counts
+int slot = SLOTXY_INV_FIRST;
+int spbslot = 0;
+
+// Menu controlled by simulating a mouse.
+bool InControlledMenu()
+{
+	return invflag || spselflag || chrflag;
+}
+
+// 0 = not near, >0 = distance related player 1 coordinates
+coords checkNearbyObjs(int x, int y, int diff)
+{
+	int diff_x = abs(plr[myplr]._px - x);
+	int diff_y = abs(plr[myplr]._py - y);
+
+	if (diff_x <= diff && diff_y <= diff) {
+		coords cm = { diff_x, diff_y };
+		//sprintf(tempstr, "N-DIFF X:%i Y:%i", diff_x, diff_y);
+		//NetSendCmdString(1 << myplr, tempstr);
+		return cm;
+	}
+	return { -1, -1 };
+}
+
+void checkItemsNearby(bool interact)
+{
+	for (int i = 0; i < MAXITEMS; i++) {
+		if (checkNearbyObjs(item[i]._ix, item[i]._iy, 1).x != -1 && item[i]._iSelFlag > 0 && item[i]._itype > -1) {
+			pcursitem = i;
+			if (dItem[item[i]._ix][item[i]._iy] <= 0)
+				continue;
+			if (interact) {
+				//sprintf(tempstr, "FOUND NEARBY ITEM AT X:%i Y:%i SEL:%i", item[i]._ix, item[i]._iy, item[i]._iSelFlag);
+				//NetSendCmdString(1 << myplr, tempstr);
+				LeftMouseCmd(false);
+			}
+			return; // item nearby, don't find objects
+		}
+	}
+	if (newCurHidden)
+		pcursitem = -1;
+	//sprintf(tempstr, "SCANNING FOR OBJECTS");
+	//NetSendCmdString(1 << myplr, tempstr);
+	for (int i = 0; i < MAXOBJECTS; i++) {
+		if (checkNearbyObjs(object[i]._ox, object[i]._oy, 1).x != -1 && object[i]._oSelFlag > 0 && object[i]._otype > -1 && currlevel) { // make sure we're in the dungeon to scan for objs
+			pcursobj = i;
+			if (interact) {
+				LeftMouseCmd(false);
+			}
+			return;
+		}
+	}
+	if (newCurHidden)
+		pcursobj = -1;
+}
+
+void checkTownersNearby(bool interact)
+{
+	for (int i = 0; i < 16; i++) {
+		if (checkNearbyObjs(towner[i]._tx, towner[i]._ty, 2).x != -1) {
+			if (towner[i]._ttype == -1)
+				continue;
+			pcursmonst = i;
+			if (interact) {
+				TalkToTowner(myplr, i);
+			}
+			break;
+		}
+	}
+}
+
+bool checkMonstersNearby(bool attack)
+{
+	int closest = 0;                 // monster ID who is closest
+	coords objDistLast = { 99, 99 }; // previous obj distance
+	// The first MAX_PLRS monsters are reserved for players' golems.
+	for (int i = MAX_PLRS; i < MAXMONSTERS; i++) {
+		int d_monster = dMonster[monster[i]._mx][monster[i]._my];
+		if (monster[i]._mFlags & MFLAG_HIDDEN || monster[i]._mhitpoints <= 0) // monster is hiding or dead, skip
+			continue;
+		if (d_monster && dFlags[monster[i]._mx][monster[i]._my] & BFLAG_LIT) {                                                                          // is monster visible

```

Comment:
This isn't taking infravisiton in to account: https://github.com/diasurgical/devilutionX/blob/master/Source/scrollrt.cpp#L541